src/CodeGen_Hexagon.cpp

```diff
@@ -175,6 +159,98 @@ Stmt sloppy_unpredicate_loads(Stmt s) {
     return SloppyUnpredicateLoads().mutate(s);
 }
 
+class InjectHVXLocks : public IRMutator2 {
+public:
+    InjectHVXLocks(const Target &t) : target(t) {
+        uses_hvx_var = Variable::make(Bool(), "uses_hvx");
+    }
+    bool uses_hvx = false;
+private:
+    Expr uses_hvx_var;
+    using IRMutator2::visit;
+    // Primarily, we do two things when we encounter a parallel for loop.
+    // First, we check if the paralell for loop uses_hvx and accordingly
+    // acqure_hvx_context i.e. acquire and release HVX locks.
+    // Then we insert a conditional unlock before the for loop, let's call
+    // this the prolog, and a conditional lock after the for loop which
+    // we shall call the epilog. So the code for a parallel loop that uses
+    // hvx should look like so.
+    //
+    // if (uses_hvx_var) {
+    //     halide_qurt_hvx_unlock();
+    // }
+    // parallel_for {
+    //     halide_qurt_hvx_lock();
+    //     ...
+    //     ...
+    //     halide_qurt_hvx_unlock();
+    // }
+    // if (uses_hvx_var) {
+    //     halide_qurt_hvx_lock();
+    // }
+    //
+    // When we move up to the enclosing scope we substitute the value of uses_hvx
+    // into the IR that should convert the conditionals to constants.
+    Stmt visit(const For *op) {
+        if (op->for_type == ForType::Parallel) {
+            bool old_uses_hvx = uses_hvx;
+            uses_hvx = false;
+
+            Stmt body = mutate(op->body);
+            Stmt s;
+            if (uses_hvx) {
+                body = acquire_hvx_context(body, target);
+                body = substitute("uses_hvx", true, body);
+                Stmt new_for = For::make(op->name, op->min, op->extent,
+                                         op->for_type, op->device_api, body);
+                Stmt prolog = IfThenElse::make(uses_hvx_var,
+                                               call_halide_qurt_hvx_unlock());
+                Stmt epilog = IfThenElse::make(uses_hvx_var,
+                                               call_halide_qurt_hvx_lock(target));
+                s = Block::make({prolog, new_for, epilog});
+                debug(4) << "Wrapping prolog & epilog around par loop\n" << s << "\n";
+            } else {
+                body = substitute("uses_hvx", false, body);
```

Comment:
I think the solution to the failure of the modified parallel_nested testcase is to not substitute here when uses_hvx is false. The logic is that it is ok to substitute uses_hvx in when it is true, because the parallel loop uses hvx in and of itself and so it will have locks and so any prologs and epilogs inside it will need to be there. However, when it is false, we cannot be so sure of substituting this value, because an enclosing scope might use_hvx and so these prologs and epilogs need to be around.

Reply:
In other words we defer the decision to an enclosing scope if we don't use_hvx. But if we do, we don't defer because that'ss all the information we need.

Reply:
That makes sense, it seems a bit conservative, but it seems like it should work, and I can't think of a less conservative thing to do quickly...

Reply:
Ok that fix works on parallel_nested.cpp (modified). Patch coming up.